File: src/main/java/com/gmail/borlandlp/minigamesdtools/nmsentities/EntityAPI.java
package com.gmail.borlandlp.minigamesdtools.nmsentities;

import net.minecraft.server.v1_12_R1.Entity;
import net.minecraft.server.v1_12_R1.EntityInsentient;
import net.minecraft.server.v1_12_R1.Vec3D;

import java.util.List;

public interface EntityAPI {
    List<PathController> getPaths();
    PathController addMovePaths(EntityInsentient entityInsentient, Vec3D[] paths, boolean isRepeating) throws Exception;
    void removeMovePath(PathController pathController);
    boolean isEntityMoveControlling(EntityInsentient entityInsentient);
    void setAttackTarget(EntityInsentient entityInsentient, EntityInsentient target);
    void setPersistentAttackTarget(EntityInsentient entityInsentient, EntityInsentient target);
    void register(String name, int id, Class<? extends Entity> oldClass, Class<? extends Entity> newClass);
    void unregister(String name, int id, Class<? extends Entity> oldClass, Class<? extends Entity> newClass);
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/nmsentities/EntityAPI.kt
package com.gmail.borlandlp.minigamesdtools.nmsentities

import com.gmail.borlandlp.minigamesdtools.nmsentities.entity.IMoveControllableEntity
import net.minecraft.server.v1_12_R1.Entity
import net.minecraft.server.v1_12_R1.EntityInsentient
import net.minecraft.server.v1_12_R1.Vec3D

interface EntityAPI {
    val paths: MutableMap<IMoveControllableEntity, PathController>
    @Throws(Exception::class)
    fun addMovePaths(
        entityInsentient: EntityInsentient,
        paths: Array<Vec3D>,
        isRepeating: Boolean
    ): PathController

    fun removeMovePath(pathController: PathController)
    fun isEntityMoveControlling(entityInsentient: EntityInsentient): Boolean
    fun setAttackTarget(entityInsentient: EntityInsentient, target: EntityInsentient)
    fun setPersistentAttackTarget(entityInsentient: EntityInsentient, target: EntityInsentient)
    fun register(
        name: String,
        id: Int,
        oldClass: Class<out Entity>,
        newClass: Class<out Entity>
    )

    fun unregister(
        name: String,
        id: Int,
        oldClass: Class<out Entity>,
        newClass: Class<out Entity>
    )
}